### PR TITLE
Add support for kafka version 3.1

### PR DESCRIPTION
**Pull request by Kuvesz**
This PR adds support for Kafka 3.1.0.
I ran the tests and it seemed to be fine, but I appreciate any insight. :)

### CODE REVIEW

File: cruise-control-metrics-reporter/src/test/java/com/linkedin/kafka/cruisecontrol/metricsreporter/utils/CCEmbeddedBroker.java [hunk 1]
```diff
@@ -38,7 +38,7 @@ public CCEmbeddedBroker(Map<Object, Object> config) {
 
     try {
       // Also validates the config
-      KafkaConfig kafkaConfig = KafkaConfig.apply(config);
+      KafkaConfig kafkaConfig = KafkaConfig.apply(config, true);
```
Comment (efeg) [inline cruise-control-metrics-reporter/src/test/java/com/linkedin/kafka/cruisecontrol/metricsreporter/utils/CCEmbeddedBroker.java:41]: Would using `new KafkaConfig(config, true)` make this line compatible with pre- and post-2.8 Kafka dependencies?

**Review by efeg — COMMENTED** [review]
Thanks for the PR! Left some comments.

-- Also, I am curious how much work would it be to keep this version upgrade a drop-in replacement with `Kafka 2.8` dependency -- i.e. would it be possible to switch back to `kafkaVersion=2.8.0` w/o code changes? For more context, please see [earlier discussion](https://github.com/linkedin/cruise-control/pull/1612#discussion_r670850521) during migration to 2.8.

File: gradle.properties [hunk 2]
```diff
@@ -2,7 +2,7 @@ org.gradle.daemon=false
 org.gradle.parallel=false
 org.gradle.jvmargs=-Xms512m -Xmx512m
 scalaVersion=2.13.6
-kafkaVersion=2.8.0
-zookeeperVersion=3.5.9
+kafkaVersion=3.1.0
```
[the remark this answers is not in the document]
Comment (Kuvesz) [reply gradle.properties:5]: It seems pretty easy to add support to 3.0 from 3.1, I'll test it quickly but I think I should first address your other comments and if all is well move on to 3.0. Would that be fine?

[the remark this answers is not in the document]
Comment (efeg) [reply gradle.properties:5]: Sure that works, thanks.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/common/KafkaNetworkClientProvider.java [hunk 3]
```diff
@@ -39,10 +46,37 @@ public NetworkClient createNetworkClient(long connectionMaxIdleMs,
...
+        time, discoverBrokerVersions, apiVersions, new LogContext());
+      } catch (NoSuchMethodException | InvocationTargetException | InstantiationException | IllegalAccessException e) {
+        LOG.debug("Unable to find Kafka 3.0- constructor for KafkaSever class", e);
+      }
+    }
+    if (networkClient != null) {
+      return networkClient;
+    } else {
+      throw new NoSuchElementException("Unable to find viable constructor for the NetworkClient class");
+    }
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/common/KafkaNetworkClientProvider.java:80]: Nit: Can we have some indentation to improve readability -- e.g.:

```
    NetworkClient networkClient = null;
    try {
      Constructor<?> kafka30PlusCon = NetworkClient.class.getConstructor(Selectable.class, Metadata.class, String.class,
                                                                         int.class, long.class, long.class, int.class, int.class, int.class,
                                                                         long.class, long.class, Time.class, boolean.class, ApiVersions.class,
                                                                         LogContext.class);
      networkClient = (NetworkClient) kafka30PlusCon.newInstance(new Selector(connectionMaxIdleMs, metrics, time, metricGrpPrefix,
                                                                              channelBuilder, new LogContext()), metadata, clientId,
                                                                 maxInFlightRequestsPerConnection, reconnectBackoffMs,
                                                                 reconnectBackoffMax, socketSendBuffer, socketReceiveBuffer,
                                                                 defaultRequestTimeoutMs, connectionSetupTimeoutMs, connectionSetupTimeoutMaxMs,
                                                                 time, discoverBrokerVersions, apiVersions, new LogContext());
    } catch (NoSuchMethodException | InvocationTargetException | InstantiationException | IllegalAccessException e) {
      LOG.debug("Unable to find Kafka 3.0+ constructor for KafkaSever class", e);
    }
    if (networkClient == null) {
      try {
        Constructor<?> kafka30MinusCon = NetworkClient.class.getConstructor(Selectable.class, Metadata.class, String.class,
                                                                            int.class, long.class, long.class, int.class, int.class, int.class,
                                                                            long.class, long.class, ClientDnsLookup.class, Time.class, boolean.class,
                                                                            ApiVersions.class, LogContext.class);
        networkClient = (NetworkClient) kafka30MinusCon.newInstance(new Selector(connectionMaxIdleMs, metrics, time, metricGrpPrefix,
                                                                                 channelBuilder, new LogContext()), metadata, clientId,
                                                                    maxInFlightRequestsPerConnection, reconnectBackoffMs,
                                                                    reconnectBackoffMax, socketSendBuffer, socketReceiveBuffer,
                                                                    defaultRequestTimeoutMs, connectionSetupTimeoutMs, connectionSetupTimeoutMaxMs,
                                                                    ClientDnsLookup.USE_ALL_DNS_IPS, time, discoverBrokerVersions, apiVersions,
                                                                    new LogContext());
      } catch (NoSuchMethodException | InvocationTargetException | InstantiationException | IllegalAccessException e) {
        LOG.debug("Unable to find Kafka 3.0- constructor for KafkaSever class", e);
      }
    }
    if (networkClient != null) {
      return networkClient;
    } else {
      throw new NoSuchElementException("Unable to find viable constructor for the NetworkClient class");
    }
```

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/KafkaCruiseControlUtils.java [hunk 4]
```diff
@@ -537,10 +538,37 @@ public static KafkaZkClient createKafkaZkClient(String connectString,
...
+        Option<String> zkClientName = Option.apply(String.format("%s-%s", metricGroup, metricType));
+        Option<ZKClientConfig> zkConfig = Option.apply(zkClientConfig);
+        Method kafka31MinusMet = KafkaZkClient.class.getMethod("apply", String.class, boolean.class, int.class,
+        int.class, int.class, org.apache.kafka.common.utils.Time.class, String.class, String.class, Option.class, Option.class);
+        kafkaZkClient = (KafkaZkClient) kafka31MinusMet.invoke(null, connectString, zkSecurityEnabled, ZK_SESSION_TIMEOUT, ZK_CONNECTION_TIMEOUT,
+        Integer.MAX_VALUE, new SystemTime(), metricGroup, metricType, zkClientName, zkConfig);
+      } catch (NoSuchMethodException | InvocationTargetException | IllegalAccessException e) {
+        LOG.debug("Unable to find apply method in KafkaZkClient for Kafka 3.1-.", e);
+      }
+    }
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/KafkaCruiseControlUtils.java:563]: Nit: Can we have some indentation to improve readability -- e.g.:

```
    try {
      String zkClientName = String.format("%s-%s", metricGroup, metricType);
      Method kafka31PlusMet = KafkaZkClient.class.getMethod("apply", String.class, boolean.class, int.class,
                                                            int.class, int.class, org.apache.kafka.common.utils.Time.class, String.class,
                                                            ZKClientConfig.class, String.class,
                                                            String.class, boolean.class);
      kafkaZkClient = (KafkaZkClient) kafka31PlusMet.invoke(null, connectString, zkSecurityEnabled, ZK_SESSION_TIMEOUT, ZK_CONNECTION_TIMEOUT,
                                                            Integer.MAX_VALUE, new SystemTime(), zkClientName, zkClientConfig, metricGroup,
                                                            metricType, false);
    } catch (NoSuchMethodException | InvocationTargetException | IllegalAccessException e) {
      LOG.debug("Unable to find apply method in KafkaZkClient for Kafka 3.1+.", e);
    }
    if (kafkaZkClient == null) {
      try {
        Option<String> zkClientName = Option.apply(String.format("%s-%s", metricGroup, metricType));
        Option<ZKClientConfig> zkConfig = Option.apply(zkClientConfig);
        Method kafka31MinusMet = KafkaZkClient.class.getMethod("apply", String.class, boolean.class, int.class,
                                                               int.class, int.class, org.apache.kafka.common.utils.Time.class, String.class,
                                                               String.class, Option.class, Option.class);
        kafkaZkClient = (KafkaZkClient) kafka31MinusMet.invoke(null, connectString, zkSecurityEnabled, ZK_SESSION_TIMEOUT, ZK_CONNECTION_TIMEOUT,
                                                               Integer.MAX_VALUE, new SystemTime(), metricGroup, metricType, zkClientName, zkConfig);
      } catch (NoSuchMethodException | InvocationTargetException | IllegalAccessException e) {
        LOG.debug("Unable to find apply method in KafkaZkClient for Kafka 3.1-.", e);
      }
    }
```

**Review by efeg — COMMENTED** [review]
Thanks for the updates and tests!
-- Left a few nits.

File: gradle.properties [hunk 5]
```diff
@@ -2,7 +2,7 @@ org.gradle.daemon=false
 org.gradle.parallel=false
 org.gradle.jvmargs=-Xms512m -Xmx512m
 scalaVersion=2.13.6
```
Comment (tomncooper) [inline gradle.properties:4]: Should we bump the Scala version to match the one used in Kafka 3.1 (Scala 2.13.8)?

File: cruise-control-metrics-reporter/src/test/java/com/linkedin/kafka/cruisecontrol/metricsreporter/utils/CCEmbeddedBroker.java [hunk 6]
```diff
@@ -38,7 +38,7 @@ public CCEmbeddedBroker(Map<Object, Object> config) {
 
     try {
       // Also validates the config
-      KafkaConfig kafkaConfig = KafkaConfig.apply(config);
+      KafkaConfig kafkaConfig = new KafkaConfig(config, true);
```
Comment (tomncooper) [inline cruise-control-metrics-reporter/src/test/java/com/linkedin/kafka/cruisecontrol/metricsreporter/utils/CCEmbeddedBroker.java:41]: you could also just use `.apply(config, true)`?

**Review by tomncooper — APPROVED** [review]
LGTM. Thanks for doing the work @Kuvesz. I got part way through doing this myself and then spotted your PR! I tested your branch with both Kafka 3.1.0 and 2.8.0 and they both build and test fine.

**Comment by tomncooper** [comment]
@efeg Ping! Are you happy with this PR?

**Review by efeg — APPROVED** [review]
@Kuvesz Thanks for the changes -- LGTM!

File: cruise-control-metrics-reporter/src/test/java/com/linkedin/kafka/cruisecontrol/metricsreporter/utils/CCEmbeddedBroker.java (hunk 6 above)
Comment (Kuvesz) [reply cruise-control-metrics-reporter/src/test/java/com/linkedin/kafka/cruisecontrol/metricsreporter/utils/CCEmbeddedBroker.java:41]: I did that originally but I think that wasn't compatible with 2.8.

File: gradle.properties (hunk 5 above)
Comment (Kuvesz) [reply gradle.properties:4]: I guess we could. Does it matter though?

**Pull request merged**